sim/prng.ts

```diff
@@ -12,21 +12,42 @@
  * @license MIT license
  */
 
+import * as sodium from 'sodium-native';
+
 /** 64-bit big-endian [high -> low] int */
 export type PRNGSeed = [number, number, number, number];
 
+interface PRNGRequired {
+	next(from?: number, to?: number): number;
+	randomChance(numerator: number, denominator: number): boolean;
+	sample<T>(items: readonly T[]): T;
+	shuffle<T>(items: T[], start: number, end: number): void;
+}
+
 /**
  * A PRNG intended to emulate the on-cartridge PRNG for Gen 5 with a 64-bit
  * initial seed.
  */
-export class PRNG {
+export class PRNG implements PRNGRequired {
 	readonly initialSeed: PRNGSeed;
-	seed: PRNGSeed;
+	seed: Buffer;
 	/** Creates a new source of randomness for the given seed. */
-	constructor(seed: PRNGSeed | null = null) {
+	constructor(seed: PRNGSeed | Buffer | null = null, initialSeed?: PRNGSeed) {
 		if (!seed) seed = PRNG.generateSeed();
-		this.initialSeed = seed.slice() as PRNGSeed; // make a copy
-		this.seed = seed.slice() as PRNGSeed;
+		// hack i know i know
+		this.initialSeed = initialSeed || [...seed] as PRNGSeed;
+		this.seed = Array.isArray(seed) ? PRNG.convertSeed(seed.slice() as PRNGSeed) : seed;
+	}
+
+	static convertSeed(inputSeed: PRNGSeed) {
+		// randombytes_buf_deterministic only takes 32 bytes (8x4 here), so it's gotta be slightly longer
+		// than we use
+		const seed = new Uint32Array(8);
+		for (let i = 0; i < seed.length; i++) seed[i] = inputSeed[i];
+		const buf = Buffer.alloc(32);
+		// @ts-ignore this doesn't accept buffers, but instead TypedArrays - typedef is out of date
+		sodium.randombytes_buf_deterministic(buf, seed);
```

Comment:
This doesn't convert the seed, it generates a new seed from the seed. It wouldn't survive serialization.